In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 
import warnings
warnings.filterwarnings("ignore")
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

In [2]:
# import data
dataset = pd.read_csv('hotel_bookings.csv')

# Data prepatation
# do not require
data = dataset.drop(['arrival_date_year', 'reservation_status', 
                  'reservation_status_date','company'], axis=1)
data

hotel  is_canceled  lead_time arrival_date_month  \
0       Resort Hotel            0        342               July   
1       Resort Hotel            0        737               July   
2       Resort Hotel            0          7               July   
3       Resort Hotel            0         13               July   
4       Resort Hotel            0         14               July   
...              ...          ...        ...                ...   
119385    City Hotel            0         23             August   
119386    City Hotel            0        102             August   
119387    City Hotel            0         34             August   
119388    City Hotel            0        109             August   
119389    City Hotel            0        205             August   

        arrival_date_week_number  arrival_date_day_of_month  \
0                             27                          1   
1                             27                          1   
2                             27                          1   
3                             27                          1   
4                             27                          1   
...                          ...                        ...   
119385                        35                         30   
119386                        35                         31   
119387                        35                         31   
119388                        35                         31   
119389                        35                         29   

        stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  \
0                             0                     0       2       0.0  ...   
1                             0                     0       2       0.0  ...   
2                             0                     1       1       0.0  ...   
3                             0                     1       1       0.0  ...   
4                             0                     2       2       0.0  ...   
...                         ...                   ...     ...       ...  ...   
119385                        2                     5       2       0.0  ...   
119386                        2                     5       3       0.0  ...   
119387                        2                     5       2       0.0  ...   
119388                        2                     5       2       0.0  ...   
119389                        2                     7       2       0.0  ...   

        reserved_room_type assigned_room_type booking_changes deposit_type  \
0                        C                  C               3   No Deposit   
1                        C                  C               4   No Deposit   
2                        A                  C               0   No Deposit   
3                        A                  A               0   No Deposit   
4                        A                  A               0   No Deposit   
...                    ...                ...             ...          ...   
119385                   A                  A               0   No Deposit   
119386                   E                  E               0   No Deposit   
119387                   D                  D               0   No Deposit   
119388                   A                  A               0   No Deposit   
119389                   A                  A               0   No Deposit   

        agent  days_in_waiting_list  customer_type     adr  \
0         NaN                     0      Transient    0.00   
1         NaN                     0      Transient    0.00   
2         NaN                     0      Transient   75.00   
3       304.0                     0      Transient   75.00   
4       240.0                     0      Transient   98.00   
...       ...                   ...            ...     ...   
119385  394.0                     0      Transient   96.14   
119386    9.0                     0      Transient  225.43 

We have 28 features, which gives us information about a customer's hotel booking and 119,390 observations, which are the number of bookings that either had or had not churned.

In [3]:
# idfentify which columns must be encoded with dummy variables
variables_to_dummy = []
for col in data.columns:
    if data[col].dtypes != 'int64':
        if data[col].dtypes != 'float64':
            variables_to_dummy.append(col)

# identify if na values
data.isna().sum()

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              4
babies                                0
meal                                  0
country                             488
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             16340
days_in_waiting_list                  0
customer_type                         0


In [4]:
data.groupby('children').size()

children
0.0     110796
1.0       4861
2.0       3652
3.0         76
10.0         1
dtype: int64

In [5]:
# The number children is not even or normally distribution 
# Therefore we will use the median to replce the na values

child_med = data['children'].median() 
data['children'] = data['children'].fillna(value=child_med)

In [6]:
# The values of the agent variable refers to a travel agency's ID
#  that made the booking. Therefore we cannot take the median or mean
#  We will input 0 for no agency.

data['agent'] = data['agent'].fillna(value=0)

data.isna().sum() # dummy variables will take care of the na values in country

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            0
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                               0
days_in_waiting_list                0
customer_type                       0
adr                                 0
required_car

In [7]:
variables_to_dummy = []
for col in data.columns:
    if data[col].dtypes == 'object':
        variables_to_dummy.append(col)


for var in variables_to_dummy:
    data[var] = pd.get_dummies(data[var])

In [8]:
for col in data.columns:
    for entry in data[col]:
        if entry < 0:
            print(entry, " in column", col) 
            # we see there is indeed a negative value in adr

-6.38  in column adr


adr stands for Average Daily Rate, which is defined by dividing the sum of all lodging transactions by the total number of staying nights. Therefore this must have been an error either when inputing the data or when the adr was calculated. Therefore we will assume the numerical value is correct and make it positive.

In [9]:
data['adr'] = data['adr'].abs()
any(data['adr'] < 0)

False

In [10]:
from sklearn.model_selection import train_test_split

X = data.drop(['is_canceled'], axis=1)
Y = data['is_canceled']
print(X.shape)
print(Y.shape)

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.50, random_state=66)

(119390, 27)
(119390,)


## Logistic Model

In [11]:
# Logistic Regression Model
from sklearn.linear_model import LogisticRegression

log_model = LogisticRegression().fit(x_train, y_train)

log_predict = log_model.predict(x_test)

In [12]:
from sklearn import metrics

log_accuracy = metrics.accuracy_score(y_test, log_predict)
log_precision = metrics.precision_score(y_test, log_predict)

print("Model's Accuracy: {}  Model's Precision: {}".format(
    round(log_accuracy,2), round(log_precision,2)))

Model's Accuracy: 0.78  Model's Precision: 0.86


In [13]:
print(metrics.classification_report(y_test, log_predict))

              precision    recall  f1-score   support

           0       0.76      0.95      0.84     37681
           1       0.86      0.48      0.61     22014

    accuracy                           0.78     59695
   macro avg       0.81      0.72      0.73     59695
weighted avg       0.80      0.78      0.76     59695



## Support Vector Machine

In [14]:
# Support Vector Machine Model

from sklearn.svm import LinearSVC

svm_model = LinearSVC(random_state = 100).fit(x_train, y_train)
svm_predict = svm_model.predict(x_test)

svm_accuracy = metrics.accuracy_score(y_test, svm_predict)
svm_precision = metrics.precision_score(y_test, svm_predict)

print("Model's Accuracy: {}  Model's Precision: {}".format(
    round(svm_accuracy,2), round(svm_precision,2)))


Model's Accuracy: 0.73  Model's Precision: 0.85


In [15]:
print(metrics.classification_report(y_test, svm_predict))

              precision    recall  f1-score   support

           0       0.71      0.97      0.82     37681
           1       0.85      0.31      0.45     22014

    accuracy                           0.73     59695
   macro avg       0.78      0.64      0.63     59695
weighted avg       0.76      0.73      0.68     59695



## Naive Bayes

In [16]:
from sklearn.naive_bayes import MultinomialNB

nb_model = MultinomialNB().fit(x_train, y_train)
nb_predict = nb_model.predict(x_test)

nb_accuracy = metrics.accuracy_score(y_test, nb_predict)
nb_precision = metrics.precision_score(y_test, nb_predict)

print("Model's Accuracy: {}  Model's Precision: {}".format(
    round(nb_accuracy,2), round(nb_precision,2)))

Model's Accuracy: 0.64  Model's Precision: 0.51


In [17]:
print(metrics.classification_report(y_test, nb_predict))

              precision    recall  f1-score   support

           0       0.73      0.69      0.71     37681
           1       0.51      0.56      0.53     22014

    accuracy                           0.64     59695
   macro avg       0.62      0.62      0.62     59695
weighted avg       0.65      0.64      0.64     59695



The naive bayes model has the least accuracy and precision. This is due to the underlying assumption that each feature is independent, hence is naive. 

## Conclusion

We will prefer to perform the the logistic regression. We were able to predict the cancellation of a booking with 78% accuracy and 86% precision.